### PR TITLE
[SPARK-10905][SparkR]: Export freqItems() for DataFrameStatFunctions

### PR DESCRIPTION
**Pull request by rerngvit**
[SPARK-10905][SparkR]: Export freqItems() for DataFrameStatFunctions
- Add function (together with roxygen2 doc) to DataFrame.R and generics.R
- Expose the function in NAMESPACE
- Add unit test for the function

### CODE REVIEW

**Comment by shivaram** [comment]
Jenkins, ok to test

**Comment by shivaram** [comment]
cc @sun-rui

File: R/pkg/R/generics.R [hunk 1]
```diff
@@ -63,6 +63,10 @@ setGeneric("countByValue", function(x) { standardGeneric("countByValue") })
 # @export
 setGeneric("crosstab", function(x, col1, col2) { standardGeneric("crosstab") })
 
+# @rdname freqItems
+# @export
+setGeneric("freqItems", function(x, ..., support = 0.01) { standardGeneric("freqItems") })
```
Comment (sun-rui) [inline R/pkg/R/generics.R]: I think using vector for cols is more R stylish.
setGeneric("freqItems", function(x, cols, support = 0.01) { standardGeneric("freqItems") })

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -1873,3 +1872,29 @@ setMethod("as.data.frame",
...
+#' @rdname freqItems
+#' @name freqItems
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
+#' fi = freqItems(df, "title", "gender")
+#' }
+setMethod("freqItems", signature(x = "DataFrame"),
+          function(x, ..., support) {
```
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: I think using vector for cols is more R stylish. 
signature(x = "DataFrame", cols = "character", support = "numeric"),
function(x, cols, support = 0.01) {

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -1873,3 +1872,29 @@ setMethod("as.data.frame",
...
           }
 )
+
+#' freqItems
+#'
+#' Finding frequent items for columns, possibly with false positives.
+#' Using the frequent element count algorithm described in
+#' \url{http://dx.doi.org/10.1145/762471.762473}, proposed by Karp, Schenker, and Papadimitriou.
+#' 
+#' @param col A list of columns to search frequent items in.
```
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: Miss the param for the DataFrame.

File: R/pkg/inst/tests/test_sparkSQL.R [hunk 4]
```diff
@@ -1322,6 +1322,19 @@ test_that("crosstab() on a DataFrame", {
...
+  rdf <- data.frame(numbers = input, letters = as.character(input),
+                    negDoubles = input * -1.0, stringsAsFactors = F)
+  rdf[ input %% 3 == 0, ] <- c(1, "1", -1)
+  df <- createDataFrame(sqlContext, rdf)
+  multiColResults <- freqItems(df, "numbers", "letters", support=0.1)
+  expect_true(1 %in% multiColResults$numbers[[1]])
+  expect_true("1" %in% multiColResults$letters[[1]])
+  singleColResult <- freqItems(df, "negDoubles", support=0.1)
+  expect_true(-1 %in% head(singleColResult$negDoubles)[[1]])
+})
```
Comment (sun-rui) [inline R/pkg/inst/tests/test_sparkSQL.R]: Could you add one more test case, which is R version of the scala test case for freqItems(), like the following:

  l <- lapply(c(0:99), function(i) { if (i %% 2 == 0) { list(1L, -1.0) } else { list(i, i * -1.0) }})
  df <- createDataFrame(sqlContext, l, c("a", "b"))
  result <- collect(freqItems(df, c("a", "b"), 0.4))
  expect_identical(result[[1]], list(list(1L, 99L)))
  expect_identical(result[[2]], list(list(-1, -99)))

**Comment by sun-rui** [comment]
@shivaram, could you merge my PR for SPARK-10752 first? that PR creates a new R file for stat functions. Then @rerngvit can rebase this PR to it.

**Comment by rerngvit** [comment]
@sun-rui I revised according to your comments. Please have a look.

File: R/pkg/R/DataFrame.R [hunk 5]
```diff
@@ -1873,3 +1872,30 @@ setMethod("as.data.frame",
...
+#' @param support (Optional) The minimum frequency for an item to be considered `frequent`. 
+#'                Should be greater than 1e-4. Default support = 0.01.
+#' @return a local R data.frame with the frequent items in each column
+#'
+#' @rdname freqItems
+#' @name freqItems
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: I think `sqlCtx` should be `sqlContext`

File: R/pkg/R/DataFrame.R [hunk 6]
```diff
@@ -1873,3 +1872,30 @@ setMethod("as.data.frame",
...
+#' fi = freqItems(df, c("title", "gender"))
+#' }
+setMethod("freqItems", signature(x = "DataFrame", cols = "character"),
+          function(x, cols, support) {
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: if support is optional shouldn't it have a default value?

Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: yes, support has a default value. support = 0.01

**Comment by sun-rui** [comment]
@rerngvit, could you rebase it to the latest master? As I moved stats functions to a new R file stats.R in https://github.com/apache/spark/pull/8869

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by rerngvit** [comment]
The errors seem not related to the PR. There might be an issue with Jenkins.
"ERROR: Timeout after 15 minutes
ERROR: Error fetching remote repo 'origin'"

**Comment by shivaram** [comment]
Jenkins, retest this please

File: R/pkg/R/DataFrame.R (hunk 5 above)
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: any update?

File: R/pkg/R/DataFrame.R (hunk 6 above)
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: any update?

File: R/pkg/inst/tests/test_sparkSQL.R [hunk 7]
```diff
@@ -1341,6 +1341,29 @@ test_that("cov() and corr() on a DataFrame", {
...
+  rdf[ input %% 3 == 0, ] <- c(1, "1", -1)
+  df <- createDataFrame(sqlContext, rdf)
+  multiColResults <- freqItems(df, c("numbers", "letters"), support=0.1)
+  expect_true(1 %in% multiColResults$numbers[[1]])
+  expect_true("1" %in% multiColResults$letters[[1]])
+  singleColResult <- freqItems(df, "negDoubles", support=0.1)
+  expect_true(-1 %in% head(singleColResult$negDoubles)[[1]])
+})
+
+test_that("freqItems2() on a DataFrame", {
```
Comment (sun-rui) [inline R/pkg/inst/tests/test_sparkSQL.R]: Don't need to add a new test_that. Just move this test case into the above one.

File: R/pkg/R/stats.R [hunk 8]
```diff
@@ -100,3 +100,30 @@ setMethod("corr",
...
+#' @rdname statfunctions
+#' @name freqItems
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
+#' fi = freqItems(df, c("title", "gender"))
+#' }
+setMethod("freqItems", signature(x = "DataFrame", cols = "character"),
+          function(x, cols, support) {
```
Comment (sun-rui) [inline R/pkg/R/stats.R]: support = 0.01

File: R/pkg/R/stats.R [hunk 9]
```diff
@@ -100,3 +100,30 @@ setMethod("corr",
...
+#' @param support (Optional) The minimum frequency for an item to be considered `frequent`. 
+#'                Should be greater than 1e-4. Default support = 0.01.
+#' @return a local R data.frame with the frequent items in each column
+#'
+#' @rdname statfunctions
+#' @name freqItems
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
```
Comment (felixcheung) [inline R/pkg/R/stats.R]: sqlCtx should be sqlContext

File: R/pkg/R/stats.R (hunk 8 above)
Comment (felixcheung) [inline R/pkg/R/stats.R]: +1

**Comment by rerngvit** [comment]
@sun-rui @felixcheung Thanks for the review. I revised according to your comments. Please have a look.

**Comment by sun-rui** [comment]
LGTM

**Comment by shivaram** [comment]
Thanks @rerngvit and @sun-rui - LGTM. Merging this

**Pull request closed without merging**